### PR TITLE
Add PID object PD limit

### PR DESCRIPTION
**Pull request by lthall**
This feature lets the user limit the maximum PD output of the PID object. This is useful in large aircraft or systems where the actuators are slew limited.

### CODE REVIEW

File: AntennaTracker/Parameters.cpp [hunk 1]
```diff
@@ -379,6 +379,14 @@ const AP_Param::Info Tracker::var_info[] = {
     // @Increment: 0.5
     // @User: Advanced
 
+    // @Param: PITCH2SRV_PDMX
+    // @DisplayName: Pitch axis PD Sum Maximum
+    // @Description: The maximum/minimum value that the sum of the P and D term can output
+    // @Range: 0 4000
+    // @Increment: 10
+    // @Units: d%
```
Comment (andyp1per) [inline AntennaTracker/Parameters.cpp:387]: Is this correct the units? You don't have it everywhere and it looks a little weird to me

**Review by IamPete1 — COMMENTED** [review]
The constrain is "invisible" because the `pid_info` object is not updated, this will make log review very difficult.

Its a little hard to say the best way to over come this. We could work out the ratio and decrease both P and D as logged. EG:
```
if (is_positive(_kpdmax)) {
    const float PD_scale = _kpdmax / MAX(fabsf(P_out + D_out), _kpdmax);
    P_out *= PD_scale;
    D_out *= PD_scale;
}
```
However doing it that way may still not make it clear as P and D will not "flat line" individually.

I think we would need that and a new limit flag. We can do a limit bitmask and in the existing `uint8_t` limit field so the log message size will not actually get bigger, although the `pid_info` struct will.

**Comment by xianglunkai** [comment]
@lthall 
I doubt the practical value of this change because based on my experience in autonomous driving, there are position, speed, and even acceleration limitations on the controller output, which is sufficient. Of course, the integration of individual saturation constraints is necessary. Perhaps my understanding is biased, thank you!

**Comment by lthall** [comment]
> I think we would need that and a new limit flag. We can do a limit bitmask and in the existing `uint8_t` limit field so the log message size will not actually get bigger, although the `pid_info` struct will.

That sounds like the best option to me.......



> @lthall I doubt the practical value of this change...

@xianglunkai this is not something that the majority of vehicles will use however it is useful for very particular problems. One of these problems is caused by slew limited actuators. Slew limited actuators have a small delay for small output changes and much larger delays for larger changes. Basically the frequency response changes with the magnitude of the signal. By adding a PD limit you can ensure the system doesn't move to the lower latency and unstable region.

**Review by bnsgeyer — APPROVED** [review]
LGTM

File: libraries/AC_PID/AC_PID.cpp [hunk 2]
```diff
@@ -172,7 +178,11 @@ float AC_PID::update_all(float target, float measurement, float dt, bool limit,
     _pid_info.P = P_out;
     _pid_info.D = D_out;
 
-    return P_out + _integrator + D_out;
+    if (is_positive(_kpdmax)) {
+        return constrain_float(P_out + D_out, -_kpdmax, _kpdmax) + _integrator;
+    } else {
+        return P_out + D_out + _integrator;
+    }
```
Comment (peterbarker) [inline libraries/AC_PID/AC_PID.cpp:185]:
```suggestion
    }
    return P_out + D_out + _integrator;
```

and elsewhere

**Comment by tridge** [comment]
@lthall I've force pushed the change we discussed to remove the constructor changes

File: AntennaTracker/Parameters.cpp (hunk 1 above)
Comment (lthall) [reply AntennaTracker/Parameters.cpp:387]: Yeh, they are using centidegrees.

**Comment by lthall** [comment]
Thanks Pete!!! I like it.

**Comment by lthall** [comment]
Mark as advanced

File: AntennaTracker/Parameters.cpp [hunk 3]
```diff
@@ -379,6 +379,14 @@ const AP_Param::Info Tracker::var_info[] = {
...
+    // @User: Standard
```
Comment (tridge) [inline AntennaTracker/Parameters.cpp:388]: need to be marked as @User: Advanced

**Review by tridge — CHANGES_REQUESTED** [review]

File: AntennaTracker/Parameters.cpp [hunk 4]
```diff
@@ -448,6 +456,14 @@ const AP_Param::Info Tracker::var_info[] = {
     // @Increment: 0.5
     // @User: Advanced
 
+    // @Param: YAW2SRV_PDMX
+    // @DisplayName: PD Sum Maximum
```
Comment (rmackay9) [inline AntennaTracker/Parameters.cpp:460]: I think this Display name should start with the axis.  So perhaps "Yaw PD Sum Maximum".  Same for other new parameter descriptions too.  Txs

**Review by tridge — APPROVED** [review]

**Pull request merged**